### ISSUE TITLE
Idea: Allow printing of coloured strings.

### ISSUE THREAD

**Issue by markusbaumeister** [gap-system/gap#2208]
It is possible to change the output colour of many terminals by using ANSI escape sequences (http://ascii-table.com/ansi-escape-sequences.php). This is already used in GAP in the file colorprompt.g to change the colours of GAP-prompt and break-loop.

For some applications (e.g. coloured graphs) it can be beneficial to colour the output of ViewObj or Display.

### Observed behaviour
For example, consider the string
```
str := "   Umbrellas: [ \033[0m( \033[34me\033[34m1\033[0m, \033[32mF\033[32m1\033[0m, \033[34me\033[34m2\033[0m, \033[32mF\033[32m2\033[0m, \033[34me\033[34m1\033[0m )\033[0m, \033[0m( \033[34me\033[34m1\033[0m,
\033[32mF\033[32m1\033[0m, \033[34me\033[34m3\033[0m, \033[32mF\033[32m2\033[0m, \033[34me\033[34m1\033[0m )\033[0m, \033[0m( \033[34me\033[34m2\033[0m, \033[32mF\033[32m1\033[0m, \033[34me\033[34m3\033[0m,
\033[32mF\033[32m2\033[0m, \033[34me\033[34m2\033[0m )\033[0m ]\n\033[0m"
```
On a screen with row length 211 (as checked by SizeScreen()) this gives the result (if we ignore the colours in the terminal):
```
   Umbrellas: [ ( e1, F1, e2, F2, e1 ), ( e1, F1, e3, F2, e1\
[0m ), ( e2, F1, e3, F2, e2 ) ]
```

Two things can be observed:
1) The string seems to be wrapped before the "actual" end of the line.
2) The string is wrapped inside one of the colour commands.
Both issues probably have the same source: GAP checks the string-length and puts breaks at appropriate places but ignores the fact that the colour commands will not take up space in the output.

### Expected behaviour
It might be possible to fix this by extending the number of special characters from 
http://www.gap-system.org/Manuals/doc/ref/chap27.html#X82E5F5AB818F32DB
(for example by also allowing ANSI escape sequences).

Since I could not localize the code where GAP handles the breaking of strings and the introduction of special characters, I can't tell how much work this change would be. I also do not know if this has further consequences down the line which I am not aware of.

**Comment by fingolfin** [comment]
I don't think this has anything to do with "special character sequences" (also known as "escaped characters" in other languages, I think).

Rather, to "fix" this, one would have to teach GAP about escape codes which have zero width when printed. Assuming one knows how to do that (I don't, though I am hopeful some googling would help), then I'd imagine it'd be a matter of messing in `src/io.c`, teaching the code in `FormatOutput` and related functions to deal with it: Right now, `TypOutputFile->pos` is used to keep track of the position on the line; this is an index into the `TypOutputFile->line` buffer. I assume this needs to be split into two variables: one which tracks the position in the line buffer, and a second which tracks the "visual" position, i.e. with zero width chars skipped (hmm, makes me wonder if this also needs to learn about non-ASCII Unicode characters? I am not just talking zero-width here, but also the fact that e.g. `ä` takes up two bytes to store, but visually just one).

That should be the majority of it, though there are some more loose ends; e.g. `GetInputLinePosition`, which is used for printing syntax warnings, and which just measure an offset inside a line buffer; this is used in `SyntaxErrorOrWarning` to print the input line, and a marker pointing at the position of an error, and assumes that every char in the input line corresponds to exactly one printed char, which is used to place a caret `^` marking the position of code causing the error/warning.
